### DISCUSSION TITLE
Building jupyterlite web site starting from a cloned repo

### DISCUSSION THREAD

**Discussion by ashok777** [jupyterlite/jupyterlite#381] [Q&A]
I would like explicit step by step instructions on how I can init, build and serve the static  jupyterlite web site beginning from a clone of https://github.com/jupyterlite/jupyterlite.git.
 
I also have a need for certain packages like pandas to be available for usage within nodebook cells.

I did read the developer guide multiple times and it was not entirely clear how I can accomplish what I am aiming  to do. 

Thanks very much

**Comment by jtpio** [comment]
Thanks @ashok777.

> I also have a need for certain packages like pandas to be available for usage within nodebook cells.

`pandas` is available in the Python kernel by default.

> I did read the developer guide multiple times and it was not entirely clear how I can accomplish what I am aiming to do.

Normally the developer guide would be the place for that. Setting up a new conda environment and running `doit` should bootstrap the repo.

Are you sure you need to start from a fork? `jupyterlite` is regularly released to PyPI so you might prefer to use a released version (more convenient): https://pypi.org/project/jupyterlite/#history

**Reply by ashok777** [reply]
Thanks for the response @jtpio.
I have a need to begin from setting up a clone of the jupyterlite forked repo. The reason is I want to embed the notebook app frame in a host app (via an iframe) . The goal is to be able to send the javascript code that was entered in a javascript notebook cell  to the host app for execution. To be able to do this, are these the steps:
1. git clone fork of jupyterlite repo.
2. Make necesary changes to javascript-kernel.
3.  From the root directory , run:  
     yarn build  
     yarn serve

Will that result in the notebook page being launched in the browser . I want to add that at this point this is just an experiment and a process od discovery.

Also when I launched jupyterlite from https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html, I could not successfully execute the following code in a python notebook cell 'import pandas as pd'. The following warning was seen:
"/lib/python3.9/site-packages/pandas/compat/__init__.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError."

So the question is how to make additional packages available for import and use in a python notebook cell. Is there a jupyterlite build time configuration involved.

**Reply by jtpio** [reply]
> The goal is to be able to send the javascript code that was entered in a javascript notebook cell to the host app for execution

As an alternative approach to the fork, you might want to have a look at writing a custom kernel extension: https://jupyterlite.readthedocs.io/en/latest/extensions.html

Which would give full control on how the code is executed in the kernel. So you might be able to put some hooks there for convenience.

> Also when I launched jupyterlite from https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html, I could not successfully execute the following code in a python notebook cell 'import pandas as pd'. The following warning was seen:
"/lib/python3.9/site-packages/pandas/compat/init.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError."

Just tested this on the latest, it takes a little while to load the library but should eventually work:

https://user-images.githubusercontent.com/591645/136851525-0987bf3f-29f7-4ad5-a053-753555028ddf.mp4



> So the question is how to make additional packages available for import and use in a python notebook cell. Is there a jupyterlite build time configuration involved.

There is some ongoing work in #310 to make that easier to work with.

**Reply by ashok777** [reply]
> As an alternative approach to the fork, you might want to have a look at writing a custom kernel extension: https://jupyterlite.readthedocs.io/en/latest/extensions.html

Agreed. A custom kernel extension is the clean way to get the desired behavior (i.e. javascript code execution occurs in the context of the host app rather than in the embedded notebook context). Wondering how this is accomplished :  Do I provide a  new implementation of the JavascriptKernel class packaged in an extension named 'javascript'
